### PR TITLE
wolfBoot as a library: configuration for use by other components

### PR DESCRIPTION
**Pull request by elms**

### CODE REVIEW

File: config/examples/library.config [hunk 1]
```diff
@@ -0,0 +1,12 @@
+ARCH=
```
Comment (danielinux) [inline config/examples/library.config:1]: Trying to compile with this configuration does not create a `target.h` file. I see there is a procedure described in lib.md. Would it be possible to have all the steps in a Makefile target? Example:

```
test-lib: include/target.h $(OBJS)
       $(Q)$(CC) $(CFLAGS) -o $@ $^
       touch empty
       $(SIGN_TOOL) --ed25519 --sha256 empty ed25519.der 1
       ./$(@) empty_v1_signed.bin`
```
And adding the following lines to `config/examples/library.c`:

 ```
+WOLFBOOT_SECTOR_SIZE=0x20000
+WOLFBOOT_PARTITION_SIZE=0x20000
```

**Review by danielinux — COMMENTED** [review]
Some minor things to fix. 
The idea is good, it needs more documentation to explain the use case.

Test target with mocks (`hal/library.c`) is very handy but should have its own test targets to work with `make distclean && cp config/examples/library.config .confg && make test-lib`

File: Makefile [hunk 2]
```diff
@@ -16,15 +16,18 @@ LDFLAGS:=
 LD_START_GROUP:=-Wl,--start-group
 LD_END_GROUP:=-Wl,--end-group
 
+V?=0
```
[the remark this answers is not in the document]
Comment (elms) [reply Makefile]: removed. It is only defined in tools/config.mk if `ARCH` is not defined.

File: tools/keytools/sign.c [hunk 3]
```diff
@@ -169,8 +169,9 @@ static void header_append_tag(uint8_t* header, uint32_t* idx, uint16_t tag,
 
 
 /* Globals */
-
+#ifdef DELTA_UPDATES
```
[the remark this answers is not in the document]
Comment (elms) [reply tools/keytools/sign.c]: Removed and added define to tools makefile to always enable it.
If this is the goal for the keytools, it should be independent of any target config. I think right now, the image header and maybe the partition size are build time set by the config.

File: config/examples/library.config (hunk 1 above)
Comment (elms) [reply config/examples/library.config:1]: ok I put a solution in place. I'll probably update the `lib.md` and squash a few commits.

**Comment by elms** [comment]
I was still not able to get the same results for the size tests so rounded up ~500 bytes for each to allow some growth.

**Comment by dgarske** [comment]
@danielinux , I rebased this and did a few minor cleanups. Can you take another pass at reviewing this? Elms has lots of great cleanups in here. Some of which I was already doing for EFI as well. Would be nice to get these merged.

**Comment by danielinux** [comment]
Thanks @dgarske - I've fixed the size check due to an increase in size of ~30B everywhere.

I'm still reviewing the build settings here, I'm still not convinced about the empty ARCH= argument in the .config for the "library" mode.

File: tools/keytools/sign.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (danielinux) [reply tools/keytools/sign.c:183]: I think it's a good compromise for now, just not having to track/map the single build options into the keytools. The image size is an important dependency, which has also been improved lately to ensure that images are created with the right manifest size.

**Review by danielinux — APPROVED** [review]

**Comment by danielinux** [comment]
Good to merge for me. Tested around and fixed a couple minor things. @dgarske please review my last commits before merging. Thanks!

**Review by danielinux — CHANGES_REQUESTED** [review]
There is a regression in autotest Group '2': SPI Flash (e.g. `test-21-forward-update-no-downgrade-SPI`)

Regression should be fixed before merging

**Review by dgarske — APPROVED** [review]

**Review by danielinux — APPROVED** [review]
Regression fixed, ready to merge

**Pull request merged**